src/app/ReadClient.h

```diff
@@ -284,6 +284,11 @@ class ReadClient : public Messaging::ExchangeDelegate
          * things like min/max intervals based on the session parameters).
          */
         virtual void OnCASESessionEstablished(const SessionHandle & aSession, ReadPrepareParams & aSubscriptionParams) {}
+
+        /*
+         * Get the last internal fatal error in callback
+         */
+        virtual CHIP_ERROR GetLastError() const { return CHIP_NO_ERROR; }
```

Comment:
Why is this new API needed?  This is making all ReadClient consumers pay extra flash costs for an internal implementation details of BufferedReadCallback (which is meant for "capable clients" only, and is better able to shoulder such costs).

If this is trying to implement the suggestion from https://github.com/project-chip/connectedhomeip/pull/30965#discussion_r1424596155, then that's not what the suggestion was.  The suggestion was that BufferedReadCallback store some internal state that will allow it to maintain the relevant invariants and just start ignoring calls from ReadClient after it reaches that error state, no?  As far as the ReadClient is concerned, there would be no changes in behavior of any sort here.

Reply:
just updated following the suggestion here.

Reply:
if we introduce GetLastError as new callback in ReadClient, we check GetLastError in the end of onAttributeData or onEventData in ReadClient, no matter what kind of chained callback we have, we could lead to Close call in readClient for read or subscribe, for resubscription case,  it would depend on which error OnResubscriptionNeeded returns, OnResubscriptionNeeded is not only conveying this error to contrloller applicaiton, but also decides how to translate the error, for example, application can skip all errors, or application can treat the errors except timeout as no error, in all situations, onError would only be called inside ReadClient, only inside Close, we may also update comments for onError, and clearly mention onError is not expected called outside ReadClient